plutus-core/untyped-plutus-core/Language/UntypedPlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -0,0 +1,568 @@
+--     then extend the environment with a binding of v to x and call computeCek on the body.
+--     If the is a builtin application then check that it's expecting a term argument,
+--     and if it's the final argument then apply the builtin to its arguments
+--     return the result, or extend the value with the new argument and call
+--     returnCek.  If v is anything else, fail.
+returnCek
+    :: (GShow uni, GEq uni, DefaultUni <: uni, Closed uni, uni `Everywhere` ExMemoryUsage)
+    => Context uni -> CekValue uni -> CekM uni (Term Name uni ())
+--- Instantiate all the free variable of the resulting term in case there are any.
+-- . ◅ V           ↦  [] V
+returnCek [] val = pure $ void $ dischargeCekValue val
+-- s , {_ A} ◅ abs α M  ↦  s ; ρ ▻ M [ α / A ]*
+returnCek (FrameForce : ctx) fun = forceEvaluate ctx fun
+-- s , [_ (M,ρ)] ◅ V  ↦  s , [V _] ; ρ ▻ M
+returnCek (FrameApplyArg argVarEnv arg : ctx) fun = do
+    computeCek (FrameApplyFun fun : ctx) argVarEnv arg
+-- s , [(lam x (M,ρ)) _] ◅ V  ↦  s ; ρ [ x  ↦  V ] ▻ M
+-- FIXME: add rule for VBuiltin once it's in the specification.
+returnCek (FrameApplyFun fun : ctx) arg = do
+    applyEvaluate ctx fun arg
+
+{- Note [Accumulating arguments].  The VBuiltin value contains lists of type and
+term arguments which grow as new arguments are encountered.  In the code below
+We just add new entries by appending to the end of the list: l -> l++[x].  This
+doesn't look terrbily good, but we don't expect the lists to ever contain more
+than three or four elements, so the cost is unlikely to be high.  We could
+accumulate lists in the normal way and reverse them when required, but this is
+error-prone and reversal adds an extra cost anyway.  We could also use something
+like Data.Sequence, but again we incur an extra cost because we have to convert
+to a normal list when passing the arguments to the constant application
+machinery.  If we really care we might want to convert the CAM to use sequences
+instead of lists.
+-}
+
+-- | @force@ a term and proceed.
+-- In case of 'VDelay' just ignore the type; for 'VBuiltin', increase
+-- the the number of @force@s, decrement the argument count,
+-- and proceed; otherwise, it's an error.
+forceEvaluate
+    :: (GShow uni, GEq uni, DefaultUni <: uni, Closed uni, uni `Everywhere` ExMemoryUsage)
+    => Context uni -> CekValue uni -> CekM uni (Term Name uni ())
+forceEvaluate ctx (VDelay _ body env) = computeCek ctx env body
+forceEvaluate ctx val@(VBuiltin ex bn arity0 arity forces args argEnv) =
+    case arity of
+      []             ->
+          throwingDischarged _MachineError EmptyBuiltinArityMachineError val
+      TermArg:_      ->
+      {- This should be impossible if we don't have zero-arity builtins:
+         we will have found this case in an earlier call to forceEvaluate
+         or applyEvaluate and called applyBuiltinName. -}
+          throwingDischarged _MachineError UnexpectedBuiltinInstantiationMachineError val'
```

Comment:
Maybe we should rename this error? It's no longer an instantiation.

Reply:
Any name suggestions? It should cover both the typed and untyped cases as we reuse all the errors.

Reply:
`BuiltinArgumentInterleavingMachineError`?  

`BuiltinTermArgumentExpectedMachineError`?  That's kind of dual to `UnexpectedBuiltinTermArgumentMachineError` which we use for the opposite case.

`TheseErrorNamesAreBecomingSoLongThatItsAlmostAsRidiculousAsJavaMachineError`?

Reply:
> `BuiltinTermArgumentExpectedMachineError`

I like the idea, thank you. Renamed the error to that.